### ISSUE TITLE
Builds can fail depending on host resources, some control is desireable

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#185]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/185)
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Jan 10, 2018, 15:14

In some scenarios, builds can fail depending on how much resources they have available. This is worsened when coupled with fault intolerant tooling, such as compilers which abort at the first sight of an OOM condition.

A recently observed case is when building LLVM (which manages to link many large binaries in parallel). In this case the build was working fine with a reasonable amount of available RAM, but begins to fail when given more CPUs, triggering more parallel linking and reaching the OOM condition.

Various approaches are possible to address the load here. A declarative approach might include some additional user configuration to describe the capabilities of the host, and / or some extension to the `.bst` format to allow one to describe the resource consumption characteristics of a given build (for instance, one might be able to express the nature of the LLVM build in a `llvm.bst` file so as to inform BuildStream about the per-job memory requirements of the element).

None of the approaches I've come up with so far are in any way attractive, though; open to suggestions.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jan 15, 2018, 06:09

changed title from **{-max-jobs is not part of the cache key: Think in a way on how to handle builds that can be build in 6 cores but not in 8, because memory exhaustion or other reasons-}** to **{+Builds can fail depending on host resources, some control is desireable+}**

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jan 15, 2018, 06:09

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Mar 9, 2019, 09:03

marked this issue as related to #942

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @abderrahimk]](https://gitlab.com/abderrahimk) on May 24, 2019, 07:59

marked this issue as related to #1033

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Oct 10, 2019, 13:21

[[Gitlab user @jjardon]](https://gitlab.com/jjardon) ping :)

What do you think needs to happen here? Did !1473 help with this?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jun 1, 2020, 08:43

marked this issue as related to #1333
